File: learn-scala-sample/src/main/scala/com/jwd/scala/sample/functional/Timer.scala
package com.jwd.scala.sample.functional

object Timer {
	def oncePerSecond(callback: () => Unit) {
		while (true) { 
		  callback(); 
		  Thread sleep 1000 
		}
	}
	
	def timeFlies() {
		println("time flies like an arrow...")
	}
	
	def main(args: Array[String]) {
		oncePerSecond(timeFlies)
		
		// Anonymous Function
		//oncePerSecond(() => println("Anonymous: time flies like an arrow..."))
	}
}

File: learn-scala-sample/src/main/scala/com/jwd/java/sample/timer/Timer.java
package com.jwd.java.sample.timer;

public class Timer {
	public void oncePerSecond(CallBack cb) {
		while (true) { 
		  cb.callback(); 
		  try {
			Thread.sleep(1000);
		  } catch (InterruptedException e) {
			  e.printStackTrace();
		  } 
		}
	}
	
	public static void main(String args[]) {
		Timer timer = new Timer();
		timer.oncePerSecond(new CallBackImpl());
		/*timer.oncePerSecond(new CallBack() {

			public void callback() {
				System.out.println("Anonymous: time flies like an arrow...");
			}
			
		});
		*/
		
		
	}
}
